src/models/geograph.py

```diff
@@ -443,27 +450,143 @@ def merge_nodes(
             and self.graph.has_node(final_index)
         ):
             raise ValueError(
-                "`final_index` must not be an existing node not in `node_list`."
+                "`final_index` cannot be an existing node that is not in `node_list`."
             )
         if final_index is None:
-            final_index = self.graph.number_of_nodes
+            final_index = self.df.last_valid_index()
 
-        adjacency_list = set()
+        # Build set of all neighbours of nodes in node_list, excluding the
+        # nodes in node_list.
+        adjacency_set = set()
         for node in node_list:
-            adjacency_list.update(self.graph.neighbors(node))
-        polygon = unary_union(
-            [self.graph.nodes[node]["geometry"] for node in node_list]
-        )
-
+            adjacency_set.update(list(self.graph.neighbors(node)))
+        adjacency_set -= adjacency_set.intersection(node_list)
+        # Build union polygon.
+        polygon = self.df.geometry.iloc[node_list].unary_union
+        # Remove nodes from graph and rows from df
         self.graph.remove_nodes_from(node_list)
+        self.df = self.df.drop(node_list)
+        # Add final node to graph and df
         self.graph.add_node(
             final_index,
             rep_point=polygon.representative_point(),
             area=polygon.area,
             perimeter=polygon.length,
-            geometry=polygon,
             class_label=class_label,
+            bounds=polygon.bounds,
+        )
+        data = {"class_label": class_label, "geometry": polygon}
+        missing_cols = {key: None for key in set(self.df.columns) - set(data.keys())}
+        self.df.loc[final_index] = gpd.GeoSeries(
+            {**data, **missing_cols}, crs=self.df.crs
         )
+        self.df = self.df.sort_index()
+        # Add edges from final node to all neighbours of the original
         self.graph.add_edges_from(
-            zip_longest([final_index], adjacency_list, fillvalue=final_index)
+            zip_longest([final_index], adjacency_set, fillvalue=final_index)
+        )
+
+    def add_habitat(
+        self,
+        name: str,
+        valid_classes: set,
+        max_travel_distance: float = 0.0,
+    ) -> None:
+        """
+        Create habitat graph and store it in habitats dictionary.
+
+        Creates a habitat subgraph of the main GeoGraph that only contains edges
+        between nodes in `valid_classes` as long as they are less than
+        `max_travel_distance` apart. All nodes which are not in `valid_classes`
+        are not in the resulting habitat graph.
+
+        Args:
+            name (str): The name of the habitat.
+            valid_classes (set): A set of class labels which make up the habitat.
+            max_travel_distance (float): The maximum distance the animal(s) in
```

Comment:
Would there be a simple way to take this argument in meters? What the unit at the moment?

Generally, it doesn't need to be perfect. Like if there is slight variations with the coordinates across a graph of like less than a few meters I think it should be fine.

Reply:
The distances are unitless at the moment to allow for passing in data which is not in metres, but I think all the distances in our data are metres anyway so this argument should already be in metres.

Reply:
OK, we decided on:
1. Converting to the passed CRS at the beginning of the `_load_from_dataframe` function
2. Adding a warning to the docstring.